src/Interpreters/getCustomKeyFilterForParallelReplicas.cpp

```diff
@@ -40,55 +40,101 @@ ASTPtr getCustomKeyFilterForParallelReplica(
+        {
+            range_upper = filter.range_upper > 0 ? RelativeSize(filter.range_upper) + RelativeSize(1)
+                                                 : RelativeSize(std::numeric_limits<UInt32>::max()) + RelativeSize(1);
+            if (range_upper > RelativeSize(std::numeric_limits<UInt32>::max()) + RelativeSize(1))
+                throw Exception(
+                    ErrorCodes::INVALID_SETTING_VALUE,
+                    "Invalid custom key range upper bound: {}. Value must be smaller than custom key column type (UInt32) max value",
+                    range_upper);
+        }
         else if (typeid_cast<const DataTypeUInt16 *>(custom_key_column_type.get()))
-            size_of_universum = RelativeSize(std::numeric_limits<UInt16>::max()) + RelativeSize(1);
+        {
+            range_upper = filter.range_upper > 0 ? RelativeSize(filter.range_upper) + RelativeSize(1)
+                                                 : RelativeSize(std::numeric_limits<UInt16>::max()) + RelativeSize(1);
+            if (range_upper > RelativeSize(std::numeric_limits<UInt16>::max()) + RelativeSize(1))
+                throw Exception(
+                    ErrorCodes::INVALID_SETTING_VALUE,
+                    "Invalid custom key range upper bound: {}. Value must be smaller than custom key column type (UInt16) max value",
+                    range_upper);
+        }
         else if (typeid_cast<const DataTypeUInt8 *>(custom_key_column_type.get()))
-            size_of_universum = RelativeSize(std::numeric_limits<UInt8>::max()) + RelativeSize(1);
+        {
+            range_upper = filter.range_upper > 0 ? RelativeSize(filter.range_upper) + RelativeSize(1)
+                                                 : RelativeSize(std::numeric_limits<UInt8>::max()) + RelativeSize(1);
+            if (range_upper > RelativeSize(std::numeric_limits<UInt8>::max()) + RelativeSize(1))
+                throw Exception(
+                    ErrorCodes::INVALID_SETTING_VALUE,
+                    "Invalid custom key range upper bound: {}. Value must be smaller than custom key column type (UInt8) max value",
+                    range_upper);
+        }
     }
 
-    if (size_of_universum == RelativeSize(0))
+    if (range_upper == RelativeSize(0))
         throw Exception(
             ErrorCodes::ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER,
             "Invalid custom key column type: {}. Must be one unsigned integer type",
             custom_key_column_type->getName());
 
+    if (range_lower >= range_upper)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Invalid custom key filter range: Range upper bound {} must be larger than range lower bound {}",
+            range_lower,
+            range_upper);
+
+    RelativeSize size_of_universum = range_upper - range_lower;
+
+    if (size_of_universum <= RelativeSize(replicas_count))
```

Comment:
can we have duplicate ranges if we have 5 replicas and range of size 6?
there is rounding when we define the boundaries so we could end up in some cases with same values for different replicas

Reply:
I don't think so, but please correct me if I misunderstood. No matter the rounding at any two sequential replicas (say replica 2 and 3) `upper_limit_rational` at replica 2 will always be the same as `lower_limit_rational` at replica 3 irrespective of the rounding. With 5 replicas, `upper_limit_rational = range_lower + ((1/5 * 2) + 1/5) * size_of_universum = range_lower + (1/5 * 3) * size_of_universum` at replica 2, which is exactly `lower_limit_rational = range_lower + (1/5 * 3) * size_of_universum` at replica 3. 

https://github.com/ClickHouse/ClickHouse/blob/6fa29dea8b4ffa97dcc749a3bde2fd9cf2320eae/src/Interpreters/getCustomKeyFilterForParallelReplicas.cpp#L117-L128

Thus the condition we add will be based on the same value at both replicas.

https://github.com/ClickHouse/ClickHouse/blob/6fa29dea8b4ffa97dcc749a3bde2fd9cf2320eae/src/Interpreters/getCustomKeyFilterForParallelReplicas.cpp#L142-L156

I added some additional tests to confirm this.